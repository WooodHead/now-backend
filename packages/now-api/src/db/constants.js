/* eslint-disable import/prefer-default-export  */

export const TABLES = {
  ACTIVITY: 'now_template',
  DEVICE: 'now_device',
  EVENT: 'now_event',
  RSVP: 'now_rsvp',
  MESSAGE: 'now_messages',
  USER: 'now_user',
};
