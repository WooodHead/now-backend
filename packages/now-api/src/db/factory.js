/* eslint-disable import/no-extraneous-dependencies  */
import { Factory } from 'rosie';
import uuid from 'uuid/v4';
import { address, name, date, internet, random, lorem, hacker } from 'faker';
import randomEmoji from 'random-emoji';
import { LocalDate, ZoneId, LocalDateTime } from 'js-joda';
import {
  EVENT_INVITE_TYPE,
  APP_INVITE_TYPE,
} from '../schema/resolvers/Invitation';

const md5 = () =>
  random
    .hexaDecimal(32)
    .toLowerCase()
    .substring(2);

Factory.define('user')
  .attr('id', uuid)
  .attr('auth0Id', () => random.alphaNumeric(24))
  .attr('email', internet.email)
  .attr('firstName', name.firstName)
  .attr('lastName', name.lastName)
  .attr('bio', lorem.paragraph)
  .attr('location', () => `${address.city()}, ${address.stateAbbr()}`)
  .attr('preferences', {})
  .attr('birthday', () =>
    LocalDate.now()
      .minusYears(20)
      .toString()
  )
  .attr('photoId', md5)
  .attr(
    'photoPreview',
    'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACgAAAAoCAYAAACM/rhtAAAAAXNSR0IArs4c6QAAAmRJREFUWAntmN1Kw0AQhU/8Fwv1F0FEKCgV9UbvvKkPIFJ8AB/C1/BZfAdF9EpBFEUEBQUp0l6opRVEi/Y4LIlx006aWHuRgTHpZHbmy9lssuh81g0dbF0dzPaNlgBGnaFEwUTBqAqY8e/vwM4OUCiYiO7Ytmfw+Bi4vRXXoUlW2wD396VhqRQGD2gL4N0dcH/fwYBGPSIWiwKq/atSsFbTlvudV60CJyduPOwU97hD7WcfH8D2NpBOA0tLwOIikM0CfX32fH/06AjgCjb29ga8vEg9E2t0bArYU8/gfod3vrcnztjcnMASeGoquMXBwe9rrMUb1lhTQBYZGwMeH91yVPXqSnx3FxgZEVgqPD8PDA5K7uWl3Jg7Us4IODvrj9p/twToL/X0BBweind3A5mMgJ6e+jPld5iFogIcH7c3skW5oG5uxG3XGQuzUFSrOAxgEJQ3Hjsgn8E4LcwU/4uCr68A348aUwHGrSDBtCqqAFMpYGBAc7/6HO1zqAJk27hVjB0waCVPTAC5HDA6qlePmdopVr0HWTBIwa0t+TYz5+EBODsTN9srxm2mVVANaFNwYcGFI8T0tPj6OvD8LKDn58D19c8NA3P/HNBxgM1NtrLb8DCwtibOHQy/y1T34gKoVIByWaB7e+3jTVStoH+KV1aAmRlTpvGxvx9gPp07I34KqWAzOFZVA3qnuKu+9vP5xlBBV6k8t2p0jalfM3wPDg1JydVVYHJSUz56jhqQragip2VjI3pjbQX1FLMgd8/Ly7JB1TaImuck/8CMKGGoZzBir5aGJ4AtyeYZlCjoEaOl0y+Le57bWRVobwAAAABJRU5ErkJggg=='
  );

Factory.define('device')
  .option('user', () => ({ id: uuid() }))
  .attr('token', uuid)
  .attr('type', () => random.arrayElement(['ios', 'android']))
  .attr('model', random.randomWord)
  .attr('userId', ['user'], ({ id }) => id);

Factory.define('rsvp')
  .option('user', () => ({ id: uuid() }))
  .option('event', () => ({ id: uuid() }))
  .attr('id', uuid)
  .attr('action', 'add')
  .attr('userId', ['user'], ({ id }) => id)
  .attr('eventId', ['event'], ({ id }) => id);

Factory.define('appInvite')
  .option('inviter', () => ({ id: uuid() }))
  .attr('id', uuid)
  .attr('type', APP_INVITE_TYPE)
  .attr('code', () => String(random.number({ max: 999999, min: 100000 })))
  .attr('expiresAt', () =>
    LocalDateTime.now()
      .plusDays(1)
      .toString()
  )
  .attr('inviterId', ['inviter'], ({ id }) => id);

Factory.define('eventInvite')
  .option('inviter', () => ({ id: uuid() }))
  .option('event', () => ({ id: uuid() }))
  .attr('id', uuid)
  .attr('type', EVENT_INVITE_TYPE)
  .attr('code', () => String(random.number({ max: 999999, min: 100000 })))
  .attr('expiresAt', () =>
    LocalDateTime.now()
      .plusDays(1)
      .toString()
  )
  .attr('inviterId', ['inviter'], ({ id }) => id)
  .attr('eventId', ['event'], ({ id }) => id);

Factory.define('activity')
  .attr('id', uuid)
  .attr('activityDate', () =>
    date
      .soon(1)
      .toISOString()
      .substring(0, 10)
  )
  .attr('title', () => `${hacker.ingverb()} for ${hacker.noun()}`)
  .attr('emoji', () => randomEmoji.random({ count: 1 })[0].character)
  .attr('description', hacker.phrase);

Factory.define('todayActivity')
  .extend('activity')
  .attr('activityDate', () => new Date().toISOString().substring(0, 10));

Factory.define('event')
  .option('activity', () => ({
    id: uuid(),
    activityDate: date
      .soon(1)
      .toISOString()
      .substring(0, 10),
  }))
  .option('location', () => ({ id: uuid() }))
  .attr('id', uuid)
  .attr('activityId', ['activity'], ({ id }) => id)
  .attr('locationId', ['location'], ({ id }) => id)
  .attr('limit', () => random.number({ max: 20, min: 5 }))
  .attr('time', ['activity'], ({ activityDate: activityDateString }) => {
    const activityDate = LocalDate.parse(activityDateString);
    return date
      .between(
        activityDate.atStartOfDay(),
        activityDate.atTime4(23, 59, 59, 999)
      )
      .toISOString();
  })
  .attr('timezone', () => random.arrayElement(ZoneId.getAvailableZoneIds()));

Factory.define('location')
  .attr('id', uuid)
  .attr('name', hacker.ingverb)
  .attr('address', address.streetAddress)
  .attr('city', address.city)
  .attr('state', address.stateAbbr)
  .attr('postalCode', address.zipCode)
  .attr('country', address.countryCode)
  .attr('neighborhood', hacker.noun)
  .attr('foursquareVenueId', md5)
  .attr(
    'location',
    () => `SRID=4326;POINT(${address.longitude()} ${address.latitude()})`
  );

Factory.define('message')
  .option('event', () => ({ id: uuid() }))
  .attr('id', uuid)
  .attr('ts', () => date.recent(1).getTime())
  .attr('text', hacker.phrase)
  .attr('eventId', ['event'], ({ id }) => id)
  .attr('userId', ['user'], ({ id }) => id);

export default Factory;
