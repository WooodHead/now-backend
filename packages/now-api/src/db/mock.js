/* eslint-disable import/no-extraneous-dependencies  */
import ApolloClient from 'apollo-client';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { SchemaLink } from 'apollo-link-schema';

import schema from '../schema';
import loaders from './loaders';

export const USER_ID = 'aea68a98-591c-11e8-9cdb-872a15ebfd30';
export const mockPromise = (resolver, rejecter) =>
  rejecter ? Promise.reject(rejecter) : Promise.resolve(resolver);

export const mocks = {
  loadMember: jest.fn(),
};

const apolloCache = new InMemoryCache();

let isAdmin = false;

export const setAdmin = b => {
  isAdmin = b;
};

export const client = new ApolloClient({
  cache: apolloCache,
  link: new SchemaLink({
    schema,
    context: () => ({
      token: null,
      user: { id: USER_ID },
      loaders: loaders({ currentUserId: USER_ID }),
      scopes: isAdmin ? ['admin'] : [],
    }),
  }),
});

export const newUserClient = (currentUserAuth0Id, currentUserId) =>
  new ApolloClient({
    cache: apolloCache,
    link: new SchemaLink({
      schema,
      context: {
        token: null,
        currentUserAuth0Id,
        loaders: loaders({ currentUserId }),
      },
    }),
  });
