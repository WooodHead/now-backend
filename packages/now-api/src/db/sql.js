import knex from 'knex';
import { setPgTypes } from './types';
import knexConfig from '../../knexfile';

const config = knexConfig[process.env.NODE_ENV];
const sql = knex(config);

setPgTypes(sql);

export default sql;
export const connection = config.connection; // eslint-disable-line prefer-destructuring

export const genRandomUuid = () => sql.raw('gen_random_uuid()');
export const makepoint = (x, y) => sql.raw('st_makepoint(?, ?)', x, y);
export const now = () => sql.raw('now()');
