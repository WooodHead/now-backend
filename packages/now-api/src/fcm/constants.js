/* eslint-disable import/prefer-default-export */
export const NEW_EVENTS_TOPIC = 'new-events';
