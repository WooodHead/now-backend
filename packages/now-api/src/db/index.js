import AWS from 'aws-sdk';
import promisify from 'util.promisify';
import { rsvpId } from '../schema/util';
import { TABLES } from './constants';
import { isDev } from '../util';

const dynamoOpts = {};
if (isDev()) {
  dynamoOpts.logger = console;
}

const dynamoDb = new AWS.DynamoDB.DocumentClient(dynamoOpts);

const pGet = promisify(dynamoDb.get.bind(dynamoDb));
const pScan = promisify(dynamoDb.scan.bind(dynamoDb));
const pUpdate = promisify(dynamoDb.update.bind(dynamoDb));
const pPut = promisify(dynamoDb.put.bind(dynamoDb));
const pQuery = promisify(dynamoDb.query.bind(dynamoDb));

export const get = (table, key = {}) =>
  pGet({
    TableName: table,
    Key: key,
  }).then(response => response.Item);

export const put = (table, item, condition = undefined) =>
  pPut({
    TableName: table,
    Item: item,
    ConditionExpression: condition,
  });

export const update = (
  table,
  key,
  expr,
  values,
  names = undefined,
  condition = undefined
) =>
  pUpdate({
    TableName: table,
    Key: key,
    ExpressionAttributeValues: values,
    ExpressionAttributeNames: names,
    UpdateExpression: expr,
    ConditionExpression: condition,
    ReturnValues: 'ALL_NEW',
  });

export const updateDynamic = (table, key, values, condition = undefined) => {
  const expressions = [];
  const expressionNames = {};
  const expressionValues = {};
  Object.entries(values).forEach(([k, v], i) => {
    const attrName = `#f${i}`;
    const fieldName = `:f${i}`;
    expressions.push(`${attrName}=${fieldName}`);
    expressionNames[attrName] = k;
    expressionValues[fieldName] = v;
  });

  return update(
    table,
    key,
    `set ${expressions.join(', ')}`,
    expressionValues,
    expressionNames,
    condition
  );
};

export const scan = (table, filter = undefined) =>
  pScan({
    TableName: table,
    FilterExpression: filter,
  }).then(response => response.Items);

export const queryRaw = params => pQuery(params);
export const query = params => pQuery(params).then(response => response.Items);

export const getUserRsvpByEvent = (userId, eventId) =>
  get(TABLES.RSVP, { id: rsvpId(eventId, userId) });

export const getActivity = id => get(TABLES.ACTIVITY, { id });

export const getEvent = id =>
  pQuery({
    TableName: TABLES.EVENT,
    KeyConditionExpression: 'id = :id',
    ExpressionAttributeValues: { ':id': id },
    IndexName: 'id-index',
  }).then(response => response.Items[0]);

export { TABLES };
