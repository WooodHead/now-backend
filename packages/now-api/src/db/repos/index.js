// @flow
import type knex from 'knex';

import defaultSql from '../sql';
import { SQL_TABLES } from '../constants';
import { putInOrder } from '../../util';

type flat = string | number;

class Repo {
  constructor(table: string, sql: knex = defaultSql) {
    this.table = table;
    this.sql = sql;
  }

  get = (query: mixed = {}) =>
    this.sql(this.table)
      .where(query)
      .first();

  all = (query: mixed = {}) => this.sql(this.table).where(query);

  insert = (obj: mixed) => this.sql(this.table).insert(obj);

  update = ({ id, ...otherFields }: { id: flat }) =>
    this.sql(this.table)
      .where({ id })
      .update(otherFields);

  batch = (ids: Array<flat>) =>
    this.sql(this.table) // TODO: log batch savings
      .whereIn('id', ids)
      .then(batch => putInOrder(batch, ids));

  delete = (query: mixed) =>
    this.sql(this.table)
      .where(query)
      .delete();

  byId = (id: flat) => this.get({ id });

  withTransaction = (trx: ?knex): Repo =>
    trx ? new Repo(this.table, trx) : this;

  sql: knex;
  table: string;
}

export const User = new Repo(SQL_TABLES.USERS);
export const Device = new Repo(SQL_TABLES.DEVICES);
export const Activity = new Repo(SQL_TABLES.ACTIVITIES);
export const Event = new Repo(SQL_TABLES.EVENTS);
export const Location = new Repo(SQL_TABLES.LOCATIONS);
export const Rsvp = new Repo(SQL_TABLES.RSVPS);
export const RsvpLog = new Repo(SQL_TABLES.RSVP_LOG);
export const Message = new Repo(SQL_TABLES.MESSAGES);
export const Invitation = new Repo(SQL_TABLES.INVITATIONS);
export const EventUserMetadata = new Repo(SQL_TABLES.EVENT_USER_METADATA);
export const ServerMessages = new Repo(SQL_TABLES.SERVER_MESSAGES);
