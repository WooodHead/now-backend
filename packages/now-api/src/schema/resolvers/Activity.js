import uuid from 'uuid';
import { LocalDateTime, LocalTime, ZoneId, LocalDate } from 'js-joda';

import { sqlPaginatify } from '../util';
import { Activity, Event } from '../../db/repos';
import sql from '../../db/sql';
import { notifyEventChange } from './Event';

export const NYC_TZ = ZoneId.of('America/New_York');
export const AVAILABILITY_HOUR = LocalTime.parse('21:00');
export const EARLY_AVAILABILITY_HOUR = LocalTime.parse('20:00');

export const getToday = () => {
  const now = LocalDateTime.now(NYC_TZ);
  return (now.toLocalTime().isBefore(AVAILABILITY_HOUR)
    ? now.toLocalDate()
    : now.toLocalDate().plusDays(1)
  ).toString();
};

const allActivities = (root, { input, orderBy = 'id' }) =>
  sqlPaginatify(orderBy, Activity.all({}), input);

const manyActivities = (root, { ids }, { loaders }) =>
  loaders.activities.loadMany(ids);

const activityQuery = (root, { id }, { loaders }) =>
  loaders.activities.load(id);

const todayActivity = () => Activity.get({ activityDate: getToday() });

export const queries = {
  todayActivity,
  activity: activityQuery,
  allActivities,
  manyActivities,
};

export const getEvents = ({ id }, { first, last, after, before }) =>
  sqlPaginatify('id', Event.all({ activityId: id }), {
    first,
    last,
    after,
    before,
  });

export const resolvers = {
  events: getEvents,
  generallyAvailableAt: () =>
    LocalDate.now()
      .atTime(AVAILABILITY_HOUR)
      .atZone(NYC_TZ),
};

const createActivity = (
  root,
  { input: { title, description, activityDate, emoji, pushNotification } },
  { loaders }
) => {
  const newId = uuid.v1();
  const newActivity = {
    id: newId,
    title,
    pushNotification,
    description,
    activityDate: activityDate.toString(),
    createdAt: sql.raw('now()'),
    updatedAt: sql.raw('now()'),
    emoji,
  };

  loaders.activities.clear(newId);

  return Activity.insert(newActivity).then(() => ({
    activity: loaders.activities.load(newId),
  }));
};

const updateActivity = async (
  root,
  { input: { id, title, description, activityDate, emoji, pushNotification } },
  { loaders }
) => {
  const updatedActivity = {
    id,
    title,
    pushNotification,
    description,
    activityDate: activityDate.toString(),
    updatedAt: sql.raw('now()'),
    emoji,
  };

  loaders.activities.clear(id);

  await Activity.update(updatedActivity);
  const activityEvents = await Event.all({ activityId: id });

  activityEvents.forEach(({ id: eventId }) => {
    notifyEventChange(eventId);
  });

  return {
    activity: loaders.activities.load(id),
  };
};

export const mutations = { createActivity, updateActivity };
