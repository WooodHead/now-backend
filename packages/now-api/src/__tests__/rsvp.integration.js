import gql from 'graphql-tag';
import { ZonedDateTime, ZoneId } from 'js-joda';

import { client, setAdmin, USER_ID } from '../db/mock';
import { SQL_TABLES } from '../db/constants';
import sql from '../db/sql';
import factory from '../db/factory';
import { Event, Rsvp, RsvpLog, Invitation } from '../db/repos';
import { NYC_TZ } from '../schema/resolvers/Activity';

const activity = factory.build('activity');
const location = factory.build('location');
const event = factory.build(
  'event',
  {
    limit: 5,
    time: ZonedDateTime.now(NYC_TZ)
      .plusHours(1)
      .withZoneSameInstant(ZoneId.UTC)
      .toString(),
    timezone: NYC_TZ.id(),
  },
  { activity, location }
);
const user = factory.build('user', { id: USER_ID });
const anotherUser = factory.build('user');

const truncateTables = () =>
  Promise.all([
    sql(SQL_TABLES.ACTIVITIES).truncate(),
    sql(SQL_TABLES.INVITATIONS).truncate(),
    sql(SQL_TABLES.INVITATION_LOG).truncate(),
    sql(SQL_TABLES.LOCATIONS).truncate(),
    sql(SQL_TABLES.EVENTS).truncate(),
    sql(SQL_TABLES.RSVPS).truncate(),
    sql(SQL_TABLES.RSVP_LOG).truncate(),
    sql(SQL_TABLES.USERS).truncate(),
  ]);

beforeEach(() =>
  truncateTables().then(() =>
    Promise.all([
      sql(SQL_TABLES.ACTIVITIES).insert(activity),
      sql(SQL_TABLES.EVENTS).insert(event),
      sql(SQL_TABLES.LOCATIONS).insert(location),
      sql(SQL_TABLES.USERS).insert([user, anotherUser]),
    ])
  ));
afterEach(() => truncateTables());

describe('Rsvp', () => {
  it('rsvp to event', async () => {
    const results = await client.mutate({
      mutation: gql`
        mutation rsvp($input: CreateRsvpInput!) {
          addRsvp(input: $input) {
            rsvp {
              id
              user {
                id
              }
              event {
                id
                isAttending
              }
            }
            event {
              id
            }
          }
        }
      `,
      variables: { input: { eventId: event.id } },
    });
    const { data } = results;
    expect(data).toMatchObject({
      addRsvp: {
        __typename: 'CreateRsvpPayload',
        event: {
          __typename: 'Event',
          id: event.id,
        },
        rsvp: {
          __typename: 'Rsvp',
          id: expect.stringMatching(
            /^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i
          ),
          event: {
            __typename: 'Event',
            id: event.id,
            isAttending: true,
          },
          user: {
            __typename: 'User',
            id: USER_ID,
          },
        },
      },
    });

    const dbRsvp = await Rsvp.byId(data.addRsvp.rsvp.id);

    expect(dbRsvp).toMatchObject({
      id: data.addRsvp.rsvp.id,
      userId: USER_ID,
      eventId: event.id,
      action: 'add',
    });

    const dbEvent = await Event.byId(event.id);

    expect(dbEvent.going).toEqual(1);
  });
  it('rsvp to event twice', async () => {
    const addRsvp = () =>
      client.mutate({
        mutation: gql`
          mutation rsvp($input: CreateRsvpInput!) {
            addRsvp(input: $input) {
              rsvp {
                id
                user {
                  id
                }
                event {
                  id
                  isAttending
                }
              }
              event {
                id
              }
            }
          }
        `,
        variables: { input: { eventId: event.id } },
      });

    await addRsvp();
    const results = await addRsvp();

    const { data } = results;
    expect(data).toMatchObject({
      addRsvp: {
        __typename: 'CreateRsvpPayload',
        event: {
          __typename: 'Event',
          id: event.id,
        },
        rsvp: {
          __typename: 'Rsvp',
          id: expect.stringMatching(
            /^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i
          ),
          event: {
            __typename: 'Event',
            id: event.id,
            isAttending: true,
          },
          user: {
            __typename: 'User',
            id: USER_ID,
          },
        },
      },
    });

    const dbRsvp = await Rsvp.byId(data.addRsvp.rsvp.id);

    expect(dbRsvp).toMatchObject({
      id: data.addRsvp.rsvp.id,
      userId: USER_ID,
      eventId: event.id,
      action: 'add',
    });

    const dbEvent = await Event.byId(event.id);

    expect(dbEvent.going).toEqual(1);
    const rsvpLog = await RsvpLog.all({ eventId: event.id, userId: USER_ID });
    expect(rsvpLog).toHaveLength(1);
    expect(rsvpLog[0]).toMatchObject({
      userId: USER_ID,
      eventId: event.id,
      action: 'add',
    });
  });

  it("Doesn't rsvp to full event", async () => {
    await Event.update({ id: event.id, going: 5 });

    const results = client.mutate({
      mutation: gql`
        mutation rsvp($input: CreateRsvpInput!) {
          addRsvp(input: $input) {
            rsvp {
              id
              user {
                id
              }
              event {
                id
                isAttending
              }
            }
            event {
              id
            }
          }
        }
      `,
      variables: { input: { eventId: event.id } },
    });
    expect.assertions(1);
    await expect(results).rejects.toEqual(
      new Error(`GraphQL error: Event ${event.id} full`)
    );
  });

  it("Doesn't rsvp to event after its start time", async () => {
    const time = ZonedDateTime.now(NYC_TZ)
      .minusMinutes(1)
      .withZoneSameInstant(ZoneId.UTC)
      .toString();

    await Event.update({ id: event.id, time });

    const results = client.mutate({
      mutation: gql`
        mutation rsvp($input: CreateRsvpInput!) {
          addRsvp(input: $input) {
            rsvp {
              id
              user {
                id
              }
              event {
                id
                isAttending
              }
            }
            event {
              id
            }
          }
        }
      `,
      variables: { input: { eventId: event.id } },
    });
    expect.assertions(1);
    await expect(results).rejects.toEqual(
      new Error(`GraphQL error: Event ${event.id} not found`)
    );
  });

  it("Doesn't rsvp to event before visibleAt", async () => {
    const visibleAt = ZonedDateTime.now(NYC_TZ)
      .plusDays(1)
      .withZoneSameInstant(ZoneId.UTC)
      .toString();

    await Event.update({ id: event.id, visibleAt });

    const results = client.mutate({
      mutation: gql`
        mutation rsvp($input: CreateRsvpInput!) {
          addRsvp(input: $input) {
            rsvp {
              id
              user {
                id
              }
              event {
                id
                isAttending
              }
            }
            event {
              id
            }
          }
        }
      `,
      variables: { input: { eventId: event.id } },
    });
    expect.assertions(1);
    await expect(results).rejects.toEqual(
      new Error(`GraphQL error: Event ${event.id} not found`)
    );
  });

  it('unrsvp from event', async () => {
    const rsvp = factory.build('rsvp', { eventId: event.id, userId: USER_ID });
    await Rsvp.insert(rsvp);
    await Event.update({ id: event.id, going: 1 });

    const results = await client.mutate({
      mutation: gql`
        mutation rsvp($input: CreateRsvpInput!) {
          removeRsvp(input: $input) {
            rsvp {
              id
            }
            event {
              id
            }
          }
        }
      `,
      variables: { input: { eventId: event.id } },
    });
    const { data } = results;
    expect(data).toMatchObject({
      removeRsvp: {
        __typename: 'CreateRsvpPayload',
        event: {
          __typename: 'Event',
          id: event.id,
        },
        rsvp: {
          __typename: 'Rsvp',
          id: rsvp.id,
        },
      },
    });

    const dbRsvp = await Rsvp.byId(rsvp.id);

    expect(dbRsvp).toMatchObject({
      id: rsvp.id,
      userId: USER_ID,
      eventId: event.id,
      action: 'remove',
    });

    const dbEvent = await Event.byId(event.id);

    expect(dbEvent.going).toEqual(0);

    const rsvpLog = await RsvpLog.all({ eventId: event.id, userId: USER_ID });
    expect(rsvpLog).toHaveLength(1);
    expect(rsvpLog[0]).toMatchObject({
      userId: USER_ID,
      eventId: event.id,
      action: 'remove',
    });
  });

  it('get user rsvps', async () => {
    const rsvp = factory.build('rsvp', {
      eventId: event.id,
      userId: USER_ID,
    });
    await Rsvp.insert(rsvp);

    const results = await client.query({
      query: gql`
        query userWithRsvps($id: ID!) {
          user(id: $id) {
            id
            rsvps {
              count
              edges {
                node {
                  id
                  lastReadTs
                }
              }
            }
          }
        }
      `,
      variables: { id: USER_ID },
    });

    const { data } = results;
    expect(data).toMatchObject({
      user: {
        __typename: 'User',
        id: USER_ID,
        rsvps: {
          count: 1,
          edges: [
            {
              node: {
                __typename: 'Rsvp',
                id: rsvp.id,
              },
            },
          ],
        },
      },
    });
  });

  it('event rsvps resolves invites', async () => {
    const invite = factory.build('eventInvite', {
      eventId: event.id,
      inviterId: USER_ID,
    });
    await Invitation.insert(invite);

    const rsvp = factory.build('rsvp', {
      eventId: event.id,
      userId: USER_ID,
    });
    await Rsvp.insert(rsvp);

    const inviteRsvp = factory.build('rsvp', {
      eventId: event.id,
      inviteId: invite.id,
    });
    await Rsvp.insert(inviteRsvp);

    const results = await client.query({
      query: gql`
        query eventRsvps($id: ID!) {
          event(id: $id) {
            rsvps {
              count
              edges {
                node {
                  id
                  user {
                    id
                  }
                  invite {
                    id
                  }
                }
              }
            }
          }
        }
      `,
      variables: { id: event.id },
    });

    const { data } = results;
    expect(data.event.rsvps).toMatchObject({
      count: 2,
      edges: expect.arrayContaining([
        expect.objectContaining({
          node: expect.objectContaining({
            id: rsvp.id,
            invite: null,
            user: expect.objectContaining({
              id: USER_ID,
            }),
          }),
        }),
        expect.objectContaining({
          node: expect.objectContaining({
            id: inviteRsvp.id,
            user: null,
            invite: expect.objectContaining({
              id: invite.id,
            }),
          }),
        }),
      ]),
    });
  });

  it('blocks non-admin users from RSVPing other users', () => {
    setAdmin(false);
    return expect(
      client.mutate({
        mutation: gql`
          mutation rsvp($input: CreateRsvpInput!) {
            addRsvp(input: $input) {
              rsvp {
                id
              }
            }
          }
        `,
        variables: { input: { eventId: event.id, userId: anotherUser.id } },
      })
    ).rejects.toMatchSnapshot();
  });

  it('allows admin users to RSVP other users', async () => {
    setAdmin(true);
    const {
      data: {
        addRsvp: { rsvp },
      },
    } = await client.mutate({
      mutation: gql`
        mutation rsvp($input: CreateRsvpInput!) {
          addRsvp(input: $input) {
            rsvp {
              id
              user {
                firstName
                id
              }
            }
          }
        }
      `,
      variables: { input: { eventId: event.id, userId: anotherUser.id } },
    });

    expect(rsvp.user.id).toEqual(anotherUser.id);
    expect(rsvp.user.firstName).toEqual(anotherUser.firstName);
  });
});
