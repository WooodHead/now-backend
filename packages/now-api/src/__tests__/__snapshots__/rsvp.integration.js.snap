// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rsvp blocks non-admin users from RSVPing other users 1`] = `[Error: GraphQL error: You can not RSVP other users.]`;
